api/envoy/config/core/v3/protocol.proto

```diff
@@ -60,15 +60,23 @@ message UpstreamHttpProtocolOptions {
       "envoy.api.v2.core.UpstreamHttpProtocolOptions";
 
   // Set transport socket `SNI <https://en.wikipedia.org/wiki/Server_Name_Indication>`_ for new
-  // upstream connections based on the downstream HTTP host/authority header, as seen by the
-  // :ref:`router filter <config_http_filters_router>`.
+  // upstream connections based on the downstream HTTP host/authority header or any other arbitrary
+  // header when :ref:`alt_header_name <envoy_v3_api_field_config.core.v3.UpstreamHttpProtocolOptions.alt_header_name>`
+  // is set, as seen by the :ref:`router filter <config_http_filters_router>`.
   bool auto_sni = 1;
 
   // Automatic validate upstream presented certificate for new upstream connections based on the
-  // downstream HTTP host/authority header, as seen by the
-  // :ref:`router filter <config_http_filters_router>`.
-  // This field is intended to set with `auto_sni` field.
+  // downstream HTTP host/authority header or any other arbitrary header when :ref:`alt_header_name <envoy_v3_api_field_config.core.v3.UpstreamHttpProtocolOptions.alt_header_name>`
+  // is set, as seen by the :ref:`router filter <config_http_filters_router>`.
+  // This field is intended to be set with `auto_sni` field.
   bool auto_san_validation = 2;
+
+  // An optional alternative to the host/authority header to be used for setting the SNI value.
+  // It should be a valid downstream HTTP header, as seen by the
+  // :ref:`router filter <config_http_filters_router>`.
+  // If unset, host/authority header will be used for populating the SNI.
+  // This field is intended to be set with `auto_sni` field.
+  string alt_header_name = 3 [(validate.rules).string = {well_known_regex: HTTP_HEADER_NAME}];
```

Comment:
What is the behavior if this field is set, but `auto_sni` is false? Document it.

What is the behavior if this field is set, and the specified header isn't present in the request? That can't happen for `:authority` but could happen here. Document it.

I think this needs a more specific name. This message could later contain other non-SNI related settings. Maybe `override_auto_sni_header`?

Reply:
add `ignore_empty: true`, eg `(validate.rules).string = {well_known_regex: HTTP_HEADER_NAME ignore_empty: true}`

Reply:
Thanks!